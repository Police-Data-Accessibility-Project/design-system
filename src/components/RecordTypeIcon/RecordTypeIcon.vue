<template>
	<FontAwesomeIcon
		v-if="recordTypesToIcons.has(recordType.toLowerCase())"
		:icon="
			recordTypesToIcons.get(recordType.toLowerCase()) ?? []
			/* Fallback array is a never case, since TS doesn't infer the typeguard from `v-if` */
		"
	/>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import {
	faPersonMilitaryToPerson,
	faPersonMilitaryPointing,
	faBuildingShield,
	faFileShield,
	faBuildingColumns,
} from '@fortawesome/free-solid-svg-icons';

/**
 * Map created from the record types taxonomy:
 * https://app.gitbook.com/o/-MXypK5ySzExtEzQU6se/s/-MXyolqTg_voOhFyAcr-/activities/data-dictionaries/record-types-taxonomy
 */
const recordTypesToIcons = ref(
	new Map([
		/***************************************************
		 ********* Police and public interactions **********
		 **************************************************/
		['police & public interactions', faPersonMilitaryToPerson], // Top level with ampersand
		['police and public interactions', faPersonMilitaryToPerson], // Top level with word
		['accident reports', faPersonMilitaryToPerson],
		['arrest records', faPersonMilitaryToPerson],
		['calls for service', faPersonMilitaryToPerson],
		['car gps', faPersonMilitaryToPerson],
		['citations', faPersonMilitaryToPerson],
		['dispatch logs', faPersonMilitaryToPerson],
		['dispatch recordings', faPersonMilitaryToPerson],
		['field contacts', faPersonMilitaryToPerson],
		['incident reports', faPersonMilitaryToPerson],
		['misc police activity', faPersonMilitaryToPerson],
		['officer involved shootings', faPersonMilitaryToPerson],
		['stops', faPersonMilitaryToPerson],
		['surveys', faPersonMilitaryToPerson],
		['use of force reports', faPersonMilitaryToPerson],
		['vehicle pursuits', faPersonMilitaryToPerson],
		/***************************************************
		 *************  Info about officers  ***************
		 **************************************************/
		['info about officers', faPersonMilitaryPointing], // Top level
		['complaints & misconduct', faPersonMilitaryPointing],
		['complaints and misconduct', faPersonMilitaryPointing],
		['daily activity logs', faPersonMilitaryPointing],
		['training & hiring info', faPersonMilitaryPointing],
		['training and hiring info', faPersonMilitaryPointing],
		['personnel records', faPersonMilitaryPointing],
		/***************************************************
		 *************  Info about agencies  ***************
		 **************************************************/
		['info about agencies', faBuildingShield], // Top level
		['annual & monthly reports', faBuildingShield],
		['annual and monthly reports', faBuildingShield],
		['budgets & finances', faBuildingShield],
		['budgets and finances', faBuildingShield],
		['contact info & agency meta', faBuildingShield],
		['contact info and agency meta', faBuildingShield],
		['geographic', faBuildingShield],
		['list of data sources', faBuildingShield],
		['policies & contracts', faBuildingShield],
		['policies and contracts', faBuildingShield],
		/***************************************************
		 *********  Agency-published resources  ************
		 **************************************************/
		['agency-published resources', faFileShield], // Top level
		['crime maps & reports', faFileShield],
		['crime maps and reports', faFileShield],
		['crime statistics', faFileShield],
		['media bulletins', faFileShield],
		['records request info', faFileShield],
		['resources', faFileShield],
		['sex offender registry', faFileShield],
		['wanted persons', faFileShield],
		/***************************************************
		 ***************  Jails & Courts  ******************
		 **************************************************/
		['jails & courts', faBuildingColumns], // Top level
		['jails and courts', faBuildingColumns], // Top level
		['jails & courts specific', faBuildingColumns], // Top level
		['jails and courts specific', faBuildingColumns], // Top level
		['booking reports', faBuildingColumns],
		['court cases', faBuildingColumns],
		['incarceration records', faBuildingColumns],
	])
);

interface Props {
	recordType: string;
}

const { recordType } = defineProps<Props>();

defineExpose({
	recordTypesToIcons,
});
</script>
