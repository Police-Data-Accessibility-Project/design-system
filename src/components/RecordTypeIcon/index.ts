export { default as RecordTypeIcon } from './RecordTypeIcon.vue';
