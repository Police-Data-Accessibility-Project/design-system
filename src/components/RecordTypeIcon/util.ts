/**
 * Creates map with type inference
 *
 * UNUSED FOR NOW, but we may need it later
 */
// export function ReadonlyMapWithStringKeys<K extends string, V>(
// 	iterable: Iterable<[K, V]>
// ): ReadonlyMap<K, V> {
// 	return new Map(iterable);
// }

// type MapKeys = typeof <<map>> extends ReadonlyMap<infer K, any>
// 	? K
// 	: never;
