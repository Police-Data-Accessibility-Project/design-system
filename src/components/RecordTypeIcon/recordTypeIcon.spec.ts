import { shallowMount } from '@vue/test-utils';
import RecordTypeIcon from './RecordTypeIcon.vue';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import {
	faPersonMilitaryToPerson,
	faPersonMilitaryPointing,
	faBuildingShield,
	faFileShield,
	faBuildingColumns,
} from '@fortawesome/free-solid-svg-icons';
import { describe, expect, it } from 'vitest';

describe('RecordTypeIcon', () => {
	it('renders the correct icon for a valid record type', () => {
		const recordType = 'Incident Reports';
		const wrapper = shallowMount(RecordTypeIcon, {
			props: { recordType },
			global: {
				stubs: {
					FontAwesomeIcon,
				},
			},
		});

		expect(wrapper.findComponent(FontAwesomeIcon).props('icon')).toEqual(
			faPersonMilitaryToPerson
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('does not render an icon for an invalid record type', () => {
		const recordType = 'Invalid Record Type';
		const wrapper = shallowMount(RecordTypeIcon, {
			props: { recordType },
			global: {
				stubs: {
					FontAwesomeIcon,
				},
			},
		});

		expect(wrapper.findComponent(FontAwesomeIcon).exists()).toBe(false);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('handles record types with ampersand correctly', () => {
		const recordType = 'Police & public interactions';
		const wrapper = shallowMount(RecordTypeIcon, {
			props: { recordType },
			global: {
				stubs: {
					FontAwesomeIcon,
				},
			},
		});

		expect(wrapper.findComponent(FontAwesomeIcon).props('icon')).toEqual(
			faPersonMilitaryToPerson
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('handles top-level record types correctly', () => {
		const recordTypes = [
			'Police & public interactions',
			'Info about officers',
			'Info about agencies',
			'Agency-published resources',
			'Jails & courts',
		];

		const expectedIcons = [
			faPersonMilitaryToPerson,
			faPersonMilitaryPointing,
			faBuildingShield,
			faFileShield,
			faBuildingColumns,
		];

		recordTypes.forEach((recordType, index) => {
			const wrapper = shallowMount(RecordTypeIcon, {
				props: { recordType },
				global: {
					stubs: {
						FontAwesomeIcon,
					},
				},
			});

			expect(wrapper.findComponent(FontAwesomeIcon).props('icon')).toEqual(
				expectedIcons[index]
			);
			expect(wrapper.html()).toMatchSnapshot();
		});
	});
});
