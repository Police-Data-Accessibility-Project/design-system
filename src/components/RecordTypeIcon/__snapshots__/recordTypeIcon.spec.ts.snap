// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecordTypeIcon > does not render an icon for an invalid record type 1`] = `<!--v-if-->`;

exports[`RecordTypeIcon > handles record types with ampersand correctly 1`] = `
<svg aria-hidden="true" class="svg-inline--fa fa-person-military-to-person" data-icon="person-military-to-person" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
  <path class="" d="M71 12.5c-8.6 1-15 8.2-15 16.8c0 9.3 7.5 16.8 16.7 16.9l111.4 0c8.8-.1 15.9-7.2 15.9-16L200 16c0-9.5-8.3-17-17.8-15.9L71 12.5zM189.5 78.1l-122.9 0C64.9 83.8 64 89.8 64 96c0 35.3 28.7 64 64 64s64-28.7 64-64c0-6.2-.9-12.2-2.5-17.9zM32 256l0 32c0 17.7 14.3 32 32 32l128 0c1.8 0 3.5-.1 5.2-.4L53 208.6C40.1 220.3 32 237.2 32 256zm190.2 42.5c1.1-3.3 1.8-6.8 1.8-10.5l0-32c0-35.3-28.7-64-64-64l-64 0c-3.7 0-7.4 .3-10.9 .9L222.2 298.5zM384 160a64 64 0 1 0 0-128 64 64 0 1 0 0 128zm-32 32c-35.3 0-64 28.7-64 64l0 32c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-32c0-35.3-28.7-64-64-64l-64 0zM215.8 450.1c5.2-4.6 8.2-11.1 8.2-18.1s-3-13.5-8.2-18.1l-64-56c-7.1-6.2-17.1-7.7-25.7-3.8S112 366.6 112 376l0 32-88 0c-13.3 0-24 10.7-24 24s10.7 24 24 24l88 0 0 32c0 9.4 5.5 18 14.1 21.9s18.6 2.4 25.7-3.8l64-56zM288 431.9c0 6.9 2.9 13.5 8.1 18.1l64 56.4c7.1 6.2 17.1 7.8 25.7 3.9s14.1-12.4 14.1-21.9l0-32.4 88 0c13.3 0 24-10.7 24-24s-10.7-24-24-24l-88 0 0-32c0-9.4-5.5-18-14.1-21.9s-18.6-2.4-25.7 3.8l-64 56c-5.2 4.5-8.2 11.1-8.2 18z" fill="currentColor" />
</svg>
`;

exports[`RecordTypeIcon > handles top-level record types correctly 1`] = `
<svg aria-hidden="true" class="svg-inline--fa fa-person-military-to-person" data-icon="person-military-to-person" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
  <path class="" d="M71 12.5c-8.6 1-15 8.2-15 16.8c0 9.3 7.5 16.8 16.7 16.9l111.4 0c8.8-.1 15.9-7.2 15.9-16L200 16c0-9.5-8.3-17-17.8-15.9L71 12.5zM189.5 78.1l-122.9 0C64.9 83.8 64 89.8 64 96c0 35.3 28.7 64 64 64s64-28.7 64-64c0-6.2-.9-12.2-2.5-17.9zM32 256l0 32c0 17.7 14.3 32 32 32l128 0c1.8 0 3.5-.1 5.2-.4L53 208.6C40.1 220.3 32 237.2 32 256zm190.2 42.5c1.1-3.3 1.8-6.8 1.8-10.5l0-32c0-35.3-28.7-64-64-64l-64 0c-3.7 0-7.4 .3-10.9 .9L222.2 298.5zM384 160a64 64 0 1 0 0-128 64 64 0 1 0 0 128zm-32 32c-35.3 0-64 28.7-64 64l0 32c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-32c0-35.3-28.7-64-64-64l-64 0zM215.8 450.1c5.2-4.6 8.2-11.1 8.2-18.1s-3-13.5-8.2-18.1l-64-56c-7.1-6.2-17.1-7.7-25.7-3.8S112 366.6 112 376l0 32-88 0c-13.3 0-24 10.7-24 24s10.7 24 24 24l88 0 0 32c0 9.4 5.5 18 14.1 21.9s18.6 2.4 25.7-3.8l64-56zM288 431.9c0 6.9 2.9 13.5 8.1 18.1l64 56.4c7.1 6.2 17.1 7.8 25.7 3.9s14.1-12.4 14.1-21.9l0-32.4 88 0c13.3 0 24-10.7 24-24s-10.7-24-24-24l-88 0 0-32c0-9.4-5.5-18-14.1-21.9s-18.6-2.4-25.7 3.8l-64 56c-5.2 4.5-8.2 11.1-8.2 18z" fill="currentColor" />
</svg>
`;

exports[`RecordTypeIcon > handles top-level record types correctly 2`] = `
<svg aria-hidden="true" class="svg-inline--fa fa-person-military-pointing" data-icon="person-military-pointing" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
  <path class="" d="M246.9 14.1C234 15.2 224 26 224 39c0 13.8 11.2 25 25 25l151 0c8.8 0 16-7.2 16-16l0-30.6C416 8 408 .7 398.7 1.4L246.9 14.1zM240 112c0 44.2 35.8 80 80 80s80-35.8 80-80c0-5.5-.6-10.8-1.6-16L241.6 96c-1 5.2-1.6 10.5-1.6 16zM72 224c-22.1 0-40 17.9-40 40s17.9 40 40 40l152 0 0 89.4L386.8 230.5c-13.3-4.3-27.3-6.5-41.6-6.5L240 224 72 224zm345.7 20.9L246.6 416 416 416l0-46.3 53.6 90.6c11.2 19 35.8 25.3 54.8 14.1s25.3-35.8 14.1-54.8L462.3 290.8c-11.2-18.9-26.6-34.5-44.6-45.9zM224 448l0 32c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-32-192 0z" fill="currentColor" />
</svg>
`;

exports[`RecordTypeIcon > handles top-level record types correctly 3`] = `
<svg aria-hidden="true" class="svg-inline--fa fa-building-shield" data-icon="building-shield" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
  <path class="" d="M0 48C0 21.5 21.5 0 48 0L336 0c26.5 0 48 21.5 48 48l0 159-42.4 17L304 224l-32 0c-8.8 0-16 7.2-16 16l0 32 0 24.2 0 7.8c0 .9 .1 1.7 .2 2.6c2.3 58.1 24.1 144.8 98.7 201.5c-5.8 2.5-12.2 3.9-18.9 3.9l-96 0 0-80c0-26.5-21.5-48-48-48s-48 21.5-48 48l0 80-96 0c-26.5 0-48-21.5-48-48L0 48zM80 224c-8.8 0-16 7.2-16 16l0 32c0 8.8 7.2 16 16 16l32 0c8.8 0 16-7.2 16-16l0-32c0-8.8-7.2-16-16-16l-32 0zm80 16l0 32c0 8.8 7.2 16 16 16l32 0c8.8 0 16-7.2 16-16l0-32c0-8.8-7.2-16-16-16l-32 0c-8.8 0-16 7.2-16 16zM64 112l0 32c0 8.8 7.2 16 16 16l32 0c8.8 0 16-7.2 16-16l0-32c0-8.8-7.2-16-16-16L80 96c-8.8 0-16 7.2-16 16zM176 96c-8.8 0-16 7.2-16 16l0 32c0 8.8 7.2 16 16 16l32 0c8.8 0 16-7.2 16-16l0-32c0-8.8-7.2-16-16-16l-32 0zm80 16l0 32c0 8.8 7.2 16 16 16l32 0c8.8 0 16-7.2 16-16l0-32c0-8.8-7.2-16-16-16l-32 0c-8.8 0-16 7.2-16 16zM423.1 225.7c5.7-2.3 12.1-2.3 17.8 0l120 48C570 277.4 576 286.2 576 296c0 63.3-25.9 168.8-134.8 214.2c-5.9 2.5-12.6 2.5-18.5 0C313.9 464.8 288 359.3 288 296c0-9.8 6-18.6 15.1-22.3l120-48zM527.4 312L432 273.8l0 187.8c68.2-33 91.5-99 95.4-149.7z" fill="currentColor" />
</svg>
`;

exports[`RecordTypeIcon > handles top-level record types correctly 4`] = `
<svg aria-hidden="true" class="svg-inline--fa fa-file-shield" data-icon="file-shield" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
  <path class="" d="M0 64C0 28.7 28.7 0 64 0L224 0l0 128c0 17.7 14.3 32 32 32l128 0 0 47-92.8 37.1c-21.3 8.5-35.2 29.1-35.2 52c0 56.6 18.9 148 94.2 208.3c-9 4.8-19.3 7.6-30.2 7.6L64 512c-35.3 0-64-28.7-64-64L0 64zm384 64l-128 0L256 0 384 128zm39.1 97.7c5.7-2.3 12.1-2.3 17.8 0l120 48C570 277.4 576 286.2 576 296c0 63.3-25.9 168.8-134.8 214.2c-5.9 2.5-12.6 2.5-18.5 0C313.9 464.8 288 359.3 288 296c0-9.8 6-18.6 15.1-22.3l120-48zM527.4 312L432 273.8l0 187.8c68.2-33 91.5-99 95.4-149.7z" fill="currentColor" />
</svg>
`;

exports[`RecordTypeIcon > renders the correct icon for a valid record type 1`] = `
<svg aria-hidden="true" class="svg-inline--fa fa-person-military-to-person" data-icon="person-military-to-person" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
  <path class="" d="M71 12.5c-8.6 1-15 8.2-15 16.8c0 9.3 7.5 16.8 16.7 16.9l111.4 0c8.8-.1 15.9-7.2 15.9-16L200 16c0-9.5-8.3-17-17.8-15.9L71 12.5zM189.5 78.1l-122.9 0C64.9 83.8 64 89.8 64 96c0 35.3 28.7 64 64 64s64-28.7 64-64c0-6.2-.9-12.2-2.5-17.9zM32 256l0 32c0 17.7 14.3 32 32 32l128 0c1.8 0 3.5-.1 5.2-.4L53 208.6C40.1 220.3 32 237.2 32 256zm190.2 42.5c1.1-3.3 1.8-6.8 1.8-10.5l0-32c0-35.3-28.7-64-64-64l-64 0c-3.7 0-7.4 .3-10.9 .9L222.2 298.5zM384 160a64 64 0 1 0 0-128 64 64 0 1 0 0 128zm-32 32c-35.3 0-64 28.7-64 64l0 32c0 17.7 14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-32c0-35.3-28.7-64-64-64l-64 0zM215.8 450.1c5.2-4.6 8.2-11.1 8.2-18.1s-3-13.5-8.2-18.1l-64-56c-7.1-6.2-17.1-7.7-25.7-3.8S112 366.6 112 376l0 32-88 0c-13.3 0-24 10.7-24 24s10.7 24 24 24l88 0 0 32c0 9.4 5.5 18 14.1 21.9s18.6 2.4 25.7-3.8l64-56zM288 431.9c0 6.9 2.9 13.5 8.1 18.1l64 56.4c7.1 6.2 17.1 7.8 25.7 3.9s14.1-12.4 14.1-21.9l0-32.4 88 0c13.3 0 24-10.7 24-24s-10.7-24-24-24l-88 0 0-32c0-9.4-5.5-18-14.1-21.9s-18.6-2.4-25.7 3.8l-64 56c-5.2 4.5-8.2 11.1-8.2 18z" fill="currentColor" />
</svg>
`;
